provider/vsphere/environ_broker.go

```diff
@@ -137,18 +139,41 @@ func (env *sessionEnviron) newRawInstance(
 	// Make sure the hostname is resolvable by adding it to /etc/hosts.
 	cloudcfg.ManageEtcHosts(true)
 
-	// If an "external network" is specified, add the boot commands
-	// necessary to configure it.
+	internalMac, err := vsphereclient.GenerateMAC()
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+
+	interfaces := []network.InterfaceInfo{{
+		InterfaceName: "eth0",
+		MACAddress:    internalMac,
+		InterfaceType: network.EthernetInterface,
+		ConfigType:    network.ConfigDHCP,
+	}}
+	networkDevices := []vsphereclient.NetworkDevice{{MAC: internalMac, Network: env.ecfg.primaryNetwork()}}
+
+	// TODO(wpk) We need to add a firewall -AND- make sure that if it's a controller we
+	// have API port open.
 	externalNetwork := env.ecfg.externalNetwork()
 	if externalNetwork != "" {
-		apiPort := 0
-		if args.InstanceConfig.Controller != nil {
-			apiPort = args.InstanceConfig.Controller.Config.APIPort()
+		externalMac, err := vsphereclient.GenerateMAC()
+		if err != nil {
+			return nil, nil, errors.Trace(err)
 		}
-		commands := common.ConfigureExternalIpAddressCommands(apiPort)
-		cloudcfg.AddBootCmd(commands...)
+		interfaces = append(interfaces, network.InterfaceInfo{
+			InterfaceName: "eth1",
+			MACAddress:    externalMac,
+			InterfaceType: network.EthernetInterface,
+			ConfigType:    network.ConfigDHCP,
+		})
+		networkDevices = append(networkDevices, vsphereclient.NetworkDevice{MAC: externalMac, Network: externalNetwork})
+	}
+	// TODO(wpk) There's no (known) way to tell cloud-init to disable network (using cloudinit.CloudInitNetworkConfigDisabled)
+	// so the network might be double-configured. That should be ok as long as we're using DHCP.
+	err = cloudcfg.AddNetworkConfig(interfaces)
```

Comment:
Is this the first place we'd error if the generated mac is not unique? I'm trying to think through how juju might generate invalid config, and make sure we catch it ourselves and deal with it, as opposed to handing off to the user.